hasher-matcher-actioner/hmalib/lambdas/api/action_rules_api.py

```diff
@@ -0,0 +1,165 @@
+            error_message = "Unexpected error."
+            handle_unexpected_error(e)
+
+        return ActionRulesAPIResponse(error_message, action_rules)
+
+    @action_rules_api.post("/", apply=[jsoninator(ActionRulesAPIRequest)])
+    def create_action_rule(
+        request: ActionRulesAPIRequest,
+    ) -> ActionRulesAPIResponse:
+        """
+        Creates an action rule.
+        """
+        logger.info("request: %s", request)
+        error_message = ""
+
+        try:
+            hmaconfig.create_config(request.action_rule)
+        except ClientError as e:
+            # TODO this test for "already exists" should be moved to a common place
+            if e.response["Error"]["Code"] == "ConditionalCheckFailedException":
+                error_message = f"An action rule with the name '{request.action_rule.name}' already exists."
+                logger.warning(
+                    "Duplicate action rule creation attempted: %s",
+                    e.response["Error"]["Message"],
+                )
+            else:
+                error_message = "Unexpected error."
+                logger.error(
+                    "Unexpected client error: %s", e.response["Error"]["Message"]
+                )
+                logger.exception(e)
+            response.status = 500
+        except Exception as e:
+            error_message = "Unexpected error."
+            handle_unexpected_error(e)
+
+        return ActionRulesAPIResponse(error_message)
+
+    @action_rules_api.put("/<old_name>", apply=[jsoninator(ActionRulesAPIRequest)])
+    def update_action_rule(
+        request: ActionRulesAPIRequest,
+        old_name: str,
+    ) -> ActionRulesAPIResponse:
+        """
+        Updates an action rule.
+        """
+        logger.info("old_name: %s", old_name)
+        logger.info("request: %s", request)
+        error_message = ""
+
+        if ActionRule.exists(request.action_rule.name):
+            try:
+                hmaconfig.update_config(request.action_rule)
+            except Exception as e:
+                error_message = "Unexpected error."
+                handle_unexpected_error(e)
+        elif ActionRule.exists(old_name):
+            try:
+                hmaconfig.create_config(request.action_rule)
+                hmaconfig.delete_config_by_type_and_name("ActionRule", old_name)
+            except Exception as e:
+                error_message = "Unexpected error."
+                handle_unexpected_error(e)
+        else:
+            error_message = f"An action rule named '{request.action_rule.name}' or '{old_name}' does not exist."
+            logger.warning(
+                "An attempt was made to update an action rule named either '%s' or '%s' but neither exist.",
+                request.action_rule.name,
+                old_name,
+            )
+            response.status = 500
+
+        return ActionRulesAPIResponse(error_message)
+
+    @action_rules_api.delete("/<name>", apply=[jsoninator])
```

Comment:
same here as for  `/update/` have the endpoint more general and have a query param for `name` ?

Reply:
I'm pretty sure DELETE HTTP requests aren't supposed to have a body. A quick google turned up this: 

From the HTTP/1.1 spec: A payload within a DELETE request message has no defined semantics; sending a payload body on a DELETE request might cause some existing implementations to reject the request.

It's pretty standard to put the ID of the thing you're deleting into the pathing. Check out the first big code block here: https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/

Reply:
right, I agree on not having a body for any GET and DELETE requests but what are your thoughts on having this be a query param instead of the endpoint path?

Reply:
A query param is still not the standard web api way to do things. We should fix the root cause of that (documented in https://github.com/facebook/ThreatExchange/issues/588 I submitted yesterday) as opposed to putting in a bunch of workarounds that we might forget to deal with.